contracts/test/BoxUpgrade.t.sol

```diff
@@ -0,0 +1,178 @@
+        assertEq(address(boxV2Proxy), address(boxV1Proxy));
+        uint256 newSize = 2;
+        uint256 newCapacity = 20;
+        boxV2Proxy.updateBox(newSize, newCapacity);
+        assertEq(boxV2Proxy.getBox().size, newSize);
+    }
+
+    // test that the ETH balance is correct after the upgrade
+    function testETHDepositCorrectWhenUpgraded() public {
+        uint256 boxSize = 1;
+        uint256 amount = 1 ether;
+        boxV1Proxy.addBox(boxSize);
+        boxV1Proxy.deposit{ value: amount }(address(this));
+        assertEq(boxV1Proxy.getBox().balance, amount);
+
+        boxV2Proxy = BoxV2(upgrader.run(address(proxy)));
+        assertEq(boxV2Proxy.getBox().balance, amount);
+    }
+
+    // test that users can withdraw even if the withdraw function
+    // did not exist in the initial implementation
+    function testIntroducingWithdrawalAfterUpgradeWorks() public {
+        vm.prank(msg.sender);
+
+        uint256 boxSize = 1;
+        uint256 amount = 1 ether;
+        boxV1Proxy.addBox(boxSize);
+
+        //deposit
+        vm.prank(msg.sender);
+        boxV1Proxy.deposit{ value: amount }(msg.sender);
+
+        vm.prank(msg.sender);
+        assertEq(boxV1Proxy.getBox().balance, amount);
+
+        //upgrade
+        boxV2Proxy = BoxV2(upgrader.run(address(proxy)));
+
+        //withdraw ETH
+        vm.prank(msg.sender);
+        uint256 userBalanceBefore = msg.sender.balance;
+
+        boxV2Proxy.withdrawETH();
+
+        //assert that the balance for that user is empty as the user withdrew their funds
+        vm.prank(msg.sender);
+
+        assertEq(boxV2Proxy.getBox().balance, 0);
+        assertEq(msg.sender.balance, userBalanceBefore + amount);
+    }
+
+    // test that overloading a method works for new implementations
+    function testNewDepositLogicWorksWithUpdatedClientPostUpgrade() public {
+        vm.prank(msg.sender);
+
+        uint256 boxSize = 1;
+        uint256 amount = 1 ether;
+        boxV1Proxy.addBox(boxSize);
+
+        //upgrade
+        boxV2Proxy = BoxV2(upgrader.run(address(proxy)));
+
+        //deposit
+        vm.prank(msg.sender);
+        boxV2Proxy.deposit{ value: amount }(address(0), msg.sender);
+
+        vm.prank(msg.sender);
+        assertEq(boxV2Proxy.getBox().balance, amount);
+    }
+
+    // test backward compatibility with an overloaded function
+    function testNewDepositLogicWorksWithOldClientPostUpgrade() public {
+        vm.prank(msg.sender);
+
+        uint256 boxSize = 1;
+        uint256 amount = 1 ether;
+        boxV1Proxy.addBox(boxSize);
+
+        //upgrade
+        boxV2Proxy = BoxV2(upgrader.run(address(proxy)));
+
+        //deposit
+        vm.prank(msg.sender);
+        boxV2Proxy.deposit{ value: amount }(msg.sender);
+
+        vm.prank(msg.sender);
+        assertEq(boxV2Proxy.getBox().balance, amount);
+    }
+
+    // test upgrading a struct works post ugprade
+    function testUpgradeNewStructElement() public {
```

Comment:
For mappings, the storage location is determined by hashing by a hash function. So the location of the struct is also determined by a hash function (instead of assigning it to the next available slot in memory). When the new variable in the struct is added, the default value is the min value for that data type, e.g. 0 for a uint. Read more: https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html